### PR TITLE
Settings page: Improve UI for post/page types to track

### LINKED ISSUE #483: Settings page: Improve UI for post/page types to track
## Is your feature request related to a problem?

We don't want users to track posts and page types twice. Right now, the UI lets them do that and we'd like to not allow them.

## Describe the solution you'd like

@GaryJones describes a good solution here: https://github.com/Parsely/wp-parsely/pull/482#issuecomment-963287549

### LINKED ISSUE #643: Consider building production CSS files
## Is your feature request related to a problem?
We generate production builds for our JS files but not for our CSS files. I'm not sure if this is by design or just something we haven't done yet.

## Describe the solution you'd like
We should consider if we want to be building our CSS files. On that same path, we could also think of including a CSS pre-processor in our pipeline which could probably do this job itself.

### PR DESCRIPTION
**Pull request by acicovic**
## Description
In the settings page, the UI for selecting the post/page types to track needs to be improved as to make it more straightforward and user friendly.

## Motivation and Context
- A more thorough discussion about this can be found in #483
- Closes #483
- Closes #643

## How Has This Been Tested?
- Manual testing of data saving and loading, including wiping the plugin's settings in the database to see how the change reacts to default/unsaved settings
- Testing that built CSS works by removing and re-enabling its wp_enqueue_style() function
- Checking on a browser that the table is responsive and works well in big and small widths
- Integration tests pass

### Before:
image

### After:
image

### CODE REVIEW

**Comment by acicovic** [comment]
Hello there,

I've just committed the UI prototype for the Track Post Types functionality, the aim being to experience the UI as well as examine the code from an accessibility perspective.

Things to know:
- I've introduced a new CSS file to hold admin CSS, but I haven't created a built file for it (yet?).
- This is a UI-only commit and I haven't tested what happens when the form saves. A saving/loading commit will come later. For the time being, you're saving at your own risk!
- Fun fact: while trying to commit this, I stumbled upon #639

Regarding accessibility
- I've replaced the usual `<legend>` that we have in most fieldsets in the settings page with a `<caption>` within the table.
- Tabbing works as expected.
- In this case, accessibility can be tricky and there are many different screen readers. I decided to use `aria-labelledby`.

Please let me know of any findings, suggestions, etc.

**Comment by acicovic** [comment]
Hi, I'm opening this for review.

SonarCloud is complaining because `class-settings-page.php` now has 21 methods from the "20 max allowed". It also suggests adding a `<legend>` in the `<fieldset>`, but I'm not sure this is needed in our case because we do have the `<caption>` within the `<table>`. If anyone has a better understanding of this, please let me know.

Thanks!

File: src/UI/class-settings-page.php [hunk 1]
```diff
@@ -285,41 +287,19 @@ function (): void {
...
-			'option_key'     => 'track_page_types',
-			'help_text'      => $h,
-			'select_options' => get_post_types( array( 'public' => true ) ),
-			'label_for'      => Parsely::OPTIONS_KEY . "[$field_id]",
-		);
-		add_settings_field(
-			'track_page_types',
-			__( 'Page Types to Track', 'wp-parsely' ),
-			array( $this, 'print_multiple_checkboxes' ),
+			__( 'Track Post Types as...', 'wp-parsely' ),
```
Comment (pauarge) [inline src/UI/class-settings-page.php]: I would remove the `...` from the label, so leave it as `Track Post Types as`

**Comment by acicovic** [comment]
Noting here some related improvements for completeness:
- ~~Build the CSS file~~ - Created as a new issue for all our CSS files in #643 
- ~~Add new UI screenshot to our WP.org page~~ - Added to #609 
 - [x] Make adjustments so the whole table is visible in very small screens (for example320px) when the post type name is really long

File: src/UI/class-settings-page.php (hunk 1 above)
Comment (acicovic) [reply src/UI/class-settings-page.php]: Not pushed yet, but done.

File: src/UI/class-settings-page.php [hunk 2]
```diff
@@ -76,23 +76,34 @@ public function run(): void {
...
 	 */
-	public function enqueue_settings_scripts( string $hook_suffix ): void {
+	public function enqueue_settings_assets( string $hook_suffix ): void {
 		if ( 'settings_page_parsely' === $hook_suffix ) {
 			wp_enqueue_media();
 
 			$admin_settings_asset = require plugin_dir_path( PARSELY_FILE ) . 'build/admin-settings.asset.php';
+			$built_assets_url     = plugin_dir_url( PARSELY_FILE ) . '/build/';
+
+			// Admin settings JS.
```
Comment (pauarge) [inline src/UI/class-settings-page.php]: I think this comment and the one in line 100 are redundant.

File: src/UI/class-settings-page.php [hunk 3]
```diff
@@ -442,41 +453,19 @@ function (): void {
 		);
 
 		// Allow use of custom taxonomy to populate articleSection in parselyPage; defaults to category.
-		$h          = __( 'By default, Parse.ly only tracks the default post type as a post page. If you want to track custom post types, select them here!', 'wp-parsely' );
```
Comment (pauarge) [inline src/UI/class-settings-page.php:445]: Could we keep this help text (or something that's similar enough but fits with the current UI)?

File: src/UI/class-settings-page.php [hunk 4]
```diff
@@ -442,41 +453,19 @@ function (): void {
...
-		$field_id   = 'track_post_types';
-		$field_args = array(
-			'title'          => __( 'Post Types to Track', 'wp-parsely' ),
-			'option_key'     => $field_id,
-			'help_text'      => $h,
-			'select_options' => get_post_types( array( 'public' => true ) ),
-			'label_for'      => Parsely::OPTIONS_KEY . "[$field_id]",
-		);
+		$field_id   = 'track_post_types_as';
+		$field_help = __( 'Select how you want to track every Post Type.', 'wp-parsely' );
```
Comment (pauarge) [inline src/UI/class-settings-page.php]: No need to capitalize `Post Type`

File: src/UI/class-settings-page.php [hunk 5]
```diff
@@ -653,7 +642,7 @@ public function print_description_text( $args ) {
 	}
 
 	/**
-	 * Print out the radio buttons.
+	 * Print out an input text tag.
```
Comment (pauarge) [inline src/UI/class-settings-page.php:643]: 👍

File: src/UI/class-settings-page.php [hunk 6]
```diff
@@ -821,6 +810,89 @@ public function print_media_single_image( array $args ) {
 		$this->print_description_text( $args );
 	}
 
+	/**
+	 * Print out the post tracking options table.
+	 *
```
Comment (pauarge) [inline src/UI/class-settings-page.php:813]: We're missing a `@since` annotation here.

File: src/UI/class-settings-page.php [hunk 7]
```diff
@@ -821,6 +810,89 @@ public function print_media_single_image( array $args ) {
...
+	 * @param array $args The arguments used in the output HTML elements.
```
Comment (pauarge) [inline src/UI/class-settings-page.php]: Would it be possible to make the `array` type more concrete here, something similar to what we do in: https://github.com/Parsely/wp-parsely/blob/89c057b5783f99b90d9d7a37659b53221eee8503/src/Integrations/class-amp.php#L158?

(if the object type is too complex to represent, feel free to dismiss my comment)

File: src/UI/class-settings-page.php [hunk 8]
```diff
@@ -821,6 +810,89 @@ public function print_media_single_image( array $args ) {
...
+
+	/**
+	 * Return the custom post type tracking values in a format that is easily
+	 * consumable by the print_track_post_types_table() function.
+	 *
+	 * @since 3.2.0
+	 *
+	 * @return array Key-value pairs with post type and their 'track as' value.
+	 */
+	public function get_tracking_values_for_display() {
```
Comment (pauarge) [inline src/UI/class-settings-page.php]: Missing return type

File: src/UI/class-settings-page.php [hunk 9]
```diff
@@ -1054,6 +1089,63 @@ public function validate_options( array $input ): array {
...
+	 *
+	 * @since 3.2.0
+	 *
+	 * @param array $input Array passed to validate_options() function.
+	 * @return void
+	 */
+	private function validate_options_post_type_tracking( array &$input ): void {
+		$options         = $this->parsely->get_options();
+		$posts           = 'track_post_types';
+		$pages           = 'track_page_types';
```
Comment (pauarge) [inline src/UI/class-settings-page.php:1104]: Maybe we could pull these two out of the function as constants?

File: src/UI/class-settings-page.php [hunk 10]
```diff
@@ -1094,7 +1186,7 @@ private static function get_logo_default(): string {
 	 * @param array $array Array of options to be sanitized.
 	 * @return array
 	 */
-	private static function validate_option_array( array $array ): array {
+	private static function sanitize_option_array( array $array ): array {
```
Comment (pauarge) [inline src/UI/class-settings-page.php:1185]: Why the name change?

**Review by pauarge — CHANGES_REQUESTED** [review]
Amazing work! Just some small tweaks left.

File: src/UI/class-settings-page.php (hunk 10 above)
Comment (acicovic) [reply src/UI/class-settings-page.php:1185]: This was done for clarity, as this function doesn't validate, but rather _**sanitizes**_ the array passed to it. As such, `sanitize` is much more descriptive of what the function does in reality, and makes it stand out from the functions starting with `validate`.

As this is a private function, changing it wasn't risky, so I went ahead with it.

File: src/UI/class-settings-page.php (hunk 7 above)
Comment (acicovic) [reply src/UI/class-settings-page.php]: So far we're only passing string key-value pairs so I've changed it to `array<string, string>`. By the way, syntax highlighting in VSCode for this isn't super great.

File: src/UI/class-settings-page.php (hunk 2 above)
Comment (acicovic) [reply src/UI/class-settings-page.php]: Removed both.

File: src/UI/class-settings-page.php (hunk 9 above)
Comment (acicovic) [reply src/UI/class-settings-page.php:1104]: I thought of this myself, but in the bigger picture I would consider this as part of #564 as an `options` class or similar which would expose constants for every option key. However, I can do this as an intermediary step if you wish.

File: src/UI/class-settings-page.php (hunk 7 above)
Comment (pauarge) [reply src/UI/class-settings-page.php]: Thanks for the change. This really helps PHPStan to infer types, in my setup (PHPStorm) works pretty well.

File: src/UI/class-settings-page.php (hunk 9 above)
Comment (pauarge) [reply src/UI/class-settings-page.php:1104]: Fair, we can leave it as is.

File: src/UI/class-settings-page.php [hunk 11]
```diff
@@ -452,7 +452,7 @@ function (): void {
 
 		// Allow use of custom taxonomy to populate articleSection in parselyPage; defaults to category.
 		$field_id   = 'track_post_types_as';
-		$field_help = __( 'Select how you want to track every post type.', 'wp-parsely' );
+		$field_help = __( 'By default, Parse.ly only tracks posts and pages. If you want to track other post types, select how you want to track them here!', 'wp-parsely' );
```
Comment (pauarge) [inline src/UI/class-settings-page.php]: Can we remove the final exclamation mark? The text looks good otherwise.

File: tests/e2e/utils.js [hunk 12]
```diff
@@ -45,3 +45,43 @@ export const checkH2DoesNotExist = async ( text ) => {
...
+	}
+
+	const [ input ] = await page.$x( '//p[contains(@class, \'submit\')]//input[contains(@name, \'screen-options-apply\')]' );
+	await input.click();
+};
+
+/**
+ * Save settings in the settings page and force a hard refresh.
+ */
+export const saveSettingsAndHardRefresh = async () => {
```
Comment (pauarge) [inline tests/e2e/utils.js:82]: Could we have similar docstring as the method above in this one?

**Review by pauarge — CHANGES_REQUESTED** [review]
Super small change, nice work on the testing!

File: src/UI/class-settings-page.php [hunk 13]
```diff
@@ -76,23 +76,32 @@ public function run(): void {
...
 		if ( 'settings_page_parsely' === $hook_suffix ) {
 			wp_enqueue_media();
 
 			$admin_settings_asset = require plugin_dir_path( PARSELY_FILE ) . 'build/admin-settings.asset.php';
+			$built_assets_url     = plugin_dir_url( PARSELY_FILE ) . '/build/';
+
 			wp_enqueue_script(
-				'parsely_admin_settings',
-				plugin_dir_url( PARSELY_FILE ) . '/build/admin-settings.js',
+				'parsely-admin-settings-js',
```
Comment (pauarge) [inline src/UI/class-settings-page.php]: Could we remove the `js` suffix to this script and the `css` to the CSS one? They'd have the same name but be different by the fact that one is a `script` enqueued and the other is a `style`.

**Review by pauarge — DISMISSED** [review]
LGTM

**Review by pauarge — APPROVED** [review]

**Pull request merged**